demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_library.py

```diff
@@ -74,48 +75,95 @@ def parse_source_paths(input_dir, valid_names):
 
         return out_file_paths
 
-    def release(self):
-        """Release internal storages"""
-
-        if self.cap is not None:
-            self.cap.release()
-
-        self.cap = None
-
     def next(self):
         """Moves pointer to the next video source"""
 
-        self.release()
+        with self.cur_source_id.get_lock():
+            self.cur_source_id.value += 1
+            if self.cur_source_id.value >= self.num_sources:
+                self.cur_source_id.value = 0
 
-        self.cur_source_id += 1
-        if self.cur_source_id >= self.num_sources:
-            self.cur_source_id = 0
+        with self.changed_source_flag.get_lock():
+            self.changed_source_flag.value = True
 
     def prev(self):
         """Moves pointer to the previous video source"""
 
-        self.release()
+        with self.cur_source_id.get_lock():
+            self.cur_source_id.value -= 1
+            if self.cur_source_id.value < 0:
+                self.cur_source_id.value = self.num_sources - 1
 
-        self.cur_source_id -= 1
-        if self.cur_source_id < 0:
-            self.cur_source_id = self.num_sources - 1
+        with self.changed_source_flag.get_lock():
+            self.changed_source_flag.value = True
 
-    def get_frame(self):
-        """Returns current frame from the active video source"""
+    def start(self):
+        """Starts internal threads"""
 
-        source_name, source_path = self.cur_source
+        if self._play_process is not None and self._play_process.is_alive():
+            return
 
-        if self.cap is None:
-            self.cap = cv2.VideoCapture(source_path)
+        self._play_process = \
+            Process(target=self._play,
+                    args=(self.visualizer, self.cur_source_id, self.source_paths,
+                          self.changed_source_flag, self.max_size, self._trg_time_step))
+        self._play_process.daemon = True
+        self._play_process.start()
 
-        _, frame = self.cap.read()
-        if frame is None:
-            self.cap.set(cv2.CAP_PROP_POS_FRAMES, 0)
-            _, frame = self.cap.read()
-            assert frame is not None
+    def release(self):
+        """Stops playing and releases internal storages"""
 
-        frame = cv2.resize(frame, self.trg_size)
-        cv2.putText(frame, 'Gesture: {}'.format(source_name), (10, frame.shape[0] - 10),
-                    cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
+        if self._play_process is not None:
+            self._play_process.terminate()
+            self._play_process.join()
 
-        return frame
+        self._play_process = None
+
+    @staticmethod
+    def _play(visualizer, cur_source_id, source_paths, changed_source_flag, max_image_size, trg_time_step):
+        """Produces live frame from the active video source"""
+
+        cap = None
+
+        start_time = time.perf_counter()
+        while True:
+            if changed_source_flag.value:
```

Comment:
Is the flag even necessary? Couldn't you store the value of `cur_source_id` in a local variable and check if `cur_source_id` changed on each iteration?

Reply:
Fixed